subsys/bluetooth/mesh/adv.c

```diff
@@ -94,24 +95,84 @@ static struct net_buf *bt_mesh_adv_create_from_pool(struct net_buf_pool *buf_poo
 	adv->type         = type;
 	adv->tag          = tag;
 	adv->xmit         = xmit;
+	adv->prio         = prio;
 
 	return buf;
 }
 
-struct net_buf *bt_mesh_adv_create(enum bt_mesh_adv_type type,
-				   enum bt_mesh_adv_tag tag,
-				   uint8_t xmit, k_timeout_t timeout)
+struct net_buf *bt_mesh_adv_relay_create(uint8_t prio, uint8_t xmit)
 {
 #if defined(CONFIG_BT_MESH_RELAY)
-	if (tag & BT_MESH_RELAY_ADV) {
-		return bt_mesh_adv_create_from_pool(&relay_buf_pool,
-						    adv_relay_pool, type,
-						    tag, xmit, timeout);
+	sys_sfnode_t *prev = NULL, *curr, *lowest_prev = NULL, *lowest = NULL;
+	uint8_t prio_cur = prio;
+	struct net_buf *buf;
+
+	buf = bt_mesh_adv_create_from_pool(&relay_buf_pool, adv_relay_pool,
+					   BT_MESH_ADV_DATA, BT_MESH_RELAY_ADV,
+					   xmit, prio, K_NO_WAIT);
+	if (buf) {
+		return buf;
+	}
+
+	if (!IS_ENABLED(CONFIG_BT_MESH_RELAY_PRIORITY) || !prio_cur) {
+		return NULL;
+	}
+
+	SYS_SFLIST_FOR_EACH_NODE(&bt_mesh_relay_queue.data_q, curr) {
+		buf = CONTAINER_OF(curr, struct net_buf, node);
+
+		if (BT_MESH_ADV(buf)->prio < prio_cur) {
+			prio_cur = BT_MESH_ADV(buf)->prio;
+			lowest_prev = prev;
+			lowest = curr;
+		}
+
+		prev = curr;
+	}
+
+	if (!lowest) {
+		return NULL;
 	}
-#endif
 
-	return bt_mesh_adv_create_from_pool(&adv_buf_pool, adv_local_pool, type,
-					    tag, xmit, timeout);
+	sys_sflist_remove(&bt_mesh_relay_queue.data_q, lowest_prev, lowest);
```

Comment:
`k_queue_remove`?

Reply:
I don't like this API, it will traverse again, obviously, we don't need additional traverse

Reply:
I did as you suggested, just because I don't want to add a kernel API,